Sources/Sentry/SentryEnvelope.m

```diff
@@ -57,10 +58,19 @@ - (instancetype)initWithHeader:(SentryEnvelopeItemHeader *)header data:(NSData *
 
 - (instancetype)initWithEvent:(SentryEvent *)event
 {
-    NSData *json = [NSJSONSerialization dataWithJSONObject:[event serialize]
-                                                   options:0
-                                                     // TODO: handle error
-                                                     error:nil];
+    NSError *error;
+    NSData *json = [SentrySerialization dataWithJSONObject:[event serialize] error:&error];
+
+    if (nil != error) {
+        SentryEvent *cantConvertEvent = [[SentryEvent alloc] initWithLevel:kSentryLevelError];
+        cantConvertEvent.message = @"Event cannot be converted to JSON.";
```

Comment:
H: I think this is a nice idea but at this level there's not much we can act if in fact this happen and a user raises an issue for us to investigate.
To make this more actionable it would require more investment here to get more information about what happened, why was this payload not a valid JSON.

A few ideas:
* Size of the object in bytes.
* Which field/part of the object not a valid JSON
* How old the event was (does it have a timestamp if its a file on disk?)
* What SDK version wrote / created the event in the first place.

Reply:
Now reading my message, I think the think we **really** need is `Which field/part of the object not a valid JSON`.
It would also be nice to know the `type` of the value which cause the issue.

That, ideally would be enough to help the developer fix the issue and avoid creating a bug report in the SDK altogether. Unless of course it was a bug on our side

Reply:
Thanks for your feedback. I like your suggestions and I agree with you that we should add more context. I totally overlooked that we would just send an event with an error message but no other context.
The problem with adding which field caused this issue is that `SentrySerialization` calls  `[NSJSONSerialization isValidJSONObject]` and we only now if the dictionary generated by `[SentryEvent serialize]` is valid or not. We don't know which property caused the problem.

Reply:
In case it's invalid, what if we traverse the object and for each of its keys we call `[NSJSONSerialization isValidJSONObject]`. If only do to the top level it'll possible be safer but very likely not find the offending field.
If we recurse to each complex we need to make sure it's not a cyclic reference.

Reply:
Yes, I agree with you that we could do that, but this seems complex to implement for me. Currently, we just drop the event in [`SentryNSUrlRequest`](https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/SentryNSURLRequest.m#L40). This is already an improvement IMO. What about doing this once we actually find out that we have many events that can't be serialized. This shouldn't happen too often anyways or do you think it's worth the effort?

Reply:
I thought of something like ([stackoverflow copy pasta incoming](https://stackoverflow.com/questions/9269372/loop-through-all-object-properties-at-runtime)):

```objective-c
objc_property_t *propertyArray = class_copyPropertyList([TestClass class], &numberOfProperties);

        for (NSUInteger i = 0; i < numberOfProperties; i++)
        {
            objc_property_t property = propertyArray[i];
            NSString *name = [[NSString alloc] initWithUTF8String:property_getName(property)];
            NSString *attributesString = [[NSString alloc] initWithUTF8String:property_getAttributes(property)];

           if ([NSJSONSerializtion isValidJSONObjecvt:[event valueForKey:name]]) {
           // Add to cantConvertEvent the name of the property that can't be serialized
           }
        }
```

If we're not going this route, I think at least we should have a kill switch (like a option to opt-out) of capturing this "warning" events.

Event though they would be nice to alert a dev that they are capturing invalid events and hence losing them, we might just spam their Sentry project and quota with these warnings without any indication of why it happens and how they can fix it. Specially if these issues are caused by our API and the developers can't find anything wrong on their usage of the SDK. They wouldn't be able to tell us anything about why this is happening besides the warning events coming in.

Final note, `cantConvertEvent.level = warning` would be a good addition.

We could add to the event already a link to the docs explaining why this happens and how it could be debugged/fixed.

`cantConvertEvent setTag:@"https://sentry.io/platforms/cocoa/invalid-json"` or something

Reply:
I think we could add something to the `SentryOptions` to disable this feature. Do we already have something similar on other SDKs?

We sanitize the properties the user adds information to the event to. The JSON conversion is only going to fail if you add garbage to [`SentryEvent.sdk`](https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/include/SentryEvent.h#L103) or [`SentryEvent.context`](https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/include/SentryEvent.h#L125) and for both we have a comment telling the user he should better know what he is doing. So this means that this should almost never happen anyway. 
Of course, we don't know if it happens often for some of our users and it could eat up their quota. At least then they would know, that they are losing events. Now they are blind and don't even know we just discard those events. I think your proposal is a great feature, but I think is not worth the effort for an edge case where a user really messes up an event with setting wrong info to properties he shouldn't touch in the first place. 

What's your opinion on this @marandaneto and @HazAT?

Reply:
I'd just drop the `sdk` and/or context if not a valid json, losing the whole event altogether is not good IMO.
Also, I think I already had this conversation with somebody and there were reasons to not raise events within the SDK, max. to add a breadcrumb with the error message (the Android SDK does it if beforeSend fails for example), but the event itself is still a customer's event.
So the user knows that something went wrong but still is not eating up their quota.

Reply:
The `sdk` shouldn't be a problem. `Contexts` is where things can fail since it's very possible someone adds something that is not serializable.

I'm fine with not adding the feature, but I'm not a fan of sending this warning events. It would be great to let the user know but if there's no indication of what and why it's broken, that's not super helpful.

If we know it's likely on the `Contexts`, we could test only that field and write the warning as: "A value set to the Context is not serializable".

Breadcrumbs is an option to avoid sending these events but it's a bit different in case of Android and .NET that we add the breadcrumb to the outgoing event that something went wrong when processing that exact event (like your callback beforeSend failed on this event. We handled it and captured the event anyway, here's a breadcrumb to let you know).

Honestly this could be as simple as:

If object isn't serializable, try to test for `contexts` only. If that's not serializable, set that to null, add a breadcrumb about it and send event.

If something else is null we could just log out and ignore. Or add a breadcrumb. Or send the event as you did but again, all of those suck because we have no idea what went wrong and what one can do to fix so maybe log and ignore is the best.